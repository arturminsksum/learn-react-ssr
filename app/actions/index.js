import { port } from '../../helpers';
const API = `http://localhost:${port}/api`;

// get articles from database
export const requestArticles = () => dispatch =>
  fetch(`${API}/articles`)
    .then(res => res.json())
    .then(articles => {
      dispatch(receiveArticles(articles));
    })
    .catch(error => console.error(error));

export const receiveArticles = articles => ({
  type: 'RECEIVE_ARTICLES',
  articles,
});

// send articles to database
export const sendPost = post =>
  fetch(`${API}/add`, {
    method: 'post',
    headers: {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(post),
  }).catch(error => console.error(error));

// export const addPost = post => ({
//   type: 'ADD_POST',
//   post,
// });

// sort articles
export const SortArticles = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_THE_VERGE: 'SHOW_THE_VERGE',
  SHOW_THE_NEXT_WEB: 'SHOW_THE_NEXT_WEB',
  SHOW_ABC_NEWS: 'SHOW_ABC_NEWS',
};

export const setSortArticles = sort => ({
  type: 'SET_SORT_ARTICLES',
  sort,
});

// login and logout
export const logIn = login => ({
  type: 'LOG_IN',
  login,
});

export const logOut = logout => ({
  type: 'LOG_OUT',
  logout,
});

export const handleLogin = (formData, history) => dispatch => {
  fetch(`${API}/login`, {
    method: 'POST',
    body: JSON.stringify(formData),
    credentials: 'same-origin',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  })
    .then(response => {
      return response.json();
    })
    .then(data => {
      if (data.error) {
        return console.error(data.error);
      }
      dispatch(logIn(true));
      return history.push('/');
    })
    .catch(error => console.error(error));
};

export const handleLogout = history => dispatch => {
  fetch(`${API}/logout`, { method: 'GET', credentials: 'same-origin' }).then(
    () => {
      dispatch(logOut(true));
      return history.push('/');
    },
  );
};

//registration
export const signUp = signup => ({
  type: 'SIGN_UP',
  signup,
});

export const handleSignup = (formData, history) => dispatch => {
  fetch(`${API}/signup`, {
    method: 'POST',
    body: JSON.stringify(formData),
    credentials: 'same-origin',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  })
    .then(response => {
      return response.json();
    })
    .then(data => {
      if (data.error) {
        return console.error(data.error);
      }
      dispatch(signUp(true));
      return history.push('/');
    })
    .catch(error => console.error(error));
};
