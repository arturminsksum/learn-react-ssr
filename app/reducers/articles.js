import defaultState from '../state.json';

const articles = (state = defaultState, action) => {
  switch (action.type) {
    case 'ADD_POST':
      return [
        ...state,
        {
          id: action.post.id,
          source: action.post.source,
          author: action.post.author,
          title: action.post.title,
          description: action.post.description,
          url: action.post.url,
        },
      ];
    default:
      return state;
  }
};
export default articles;
