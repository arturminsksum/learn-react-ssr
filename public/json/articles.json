[
  {
    "id": "the-verge",
    "source": "The Verge",
    "author": "Thuy Ong",
    "title": "Facebook begins privacy push ahead of tough new European law",
    "description":
      "Facebook’s new privacy center tool is a response the upcoming General Data Protection Regulation which will be introduced in May",
    "url":
      "https://www.theverge.com/2018/1/29/16944304/facebook-privacy-eu-law-general-data-protection-regulation",
    "urlToImage":
      "https://cdn.vox-cdn.com/thumbor/FQYYmGkfUAzxuJyhX3di-vw9soQ=/0x146:2040x1214/fit-in/1200x630/cdn.vox-cdn.com/uploads/chorus_asset/file/3446688/facebook-stock-1102.0.jpg",
    "publishedAt": "2018-01-29T12:10:18Z"
  },
  {
    "id": "the-next-web",
    "source": "The Next Web",
    "author": "Mix",
    "title": "We got a cryptocurrency mystery box – but you probably shouldn't",
    "description":
      "Earlier in January we wrote about a quirky new trend: cryptocurrency mystery boxes you can order online to get a random selection of digital coins.\r\n\r\nCBlocks – the company we spoke ...",
    "url":
      "https://thenextweb.com/hardfork/2018/01/29/cryptocurrency-mystery-box-review/",
    "urlToImage":
      "https://cdn0.tnwcdn.com/wp-content/blogs.dir/1/files/2018/01/cryptocurrency-1.png",
    "publishedAt": "2018-01-29T11:49:06Z"
  },
  {
    "id": "abc-news",
    "source": "ABC News",
    "author": "ABC",
    "title": "Rousey's WWE deal gets ex-UFC star back in bright spotlight",
    "description":
      "Ronda Rousey stayed silent as she pointed from the ring toward the WrestleMania logo that hung in the rafters. Without saying a word, the MMA great made clear she was signaling her wrestling debut at WWE's signature event. Rousey wants her WrestleMania moment…",
    "url":
      "http://abcnews.go.com/Entertainment/wireStory/rouseys-wwe-deal-ufc-star-back-bright-spotlight-52691202",
    "urlToImage":
      "http://a.abcnews.com/images/Entertainment/WireAP_18f9d8b3c72e4d94a0cc6332ee8924cd_16x9_992.jpg",
    "publishedAt": "2018-01-29T22:01:09Z"
  }
]
