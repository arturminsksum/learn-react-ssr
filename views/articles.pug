extends layout

block content

  for item in articles
    article.media.box
      figure.media-left
        p.image.is-64x64
          img(src=item.urlToImage)
      .media-content
        .content
          a.title(href=item.url)=item.title
          p
            span=item.description
          p
            strong Source: #{item.source}
            br
            small Author: #{item.author}
            br
            small Time: #{item.publishedAt}

  .columns
    .column.is-one-third
      form(method='post')
        .field
          label.label Subject
          .control
            .select
              select(name="id")
                option(value='the-verge') The Verge
                option(value='the-next-web') The Next Web
                option(value='abc-news' selected) ABC News
        .field
          label.label Source
          .control
            input.input(type='text', placeholder='Source', name="source", value='ABC News')
        .field
          label.label Author
          .control
            input.input(type='text', placeholder='Author', name="author", value='ABC News')
        .field
          label.label Title
          .control
            input.input(type='text', placeholder='Title', name="title", value='WATCH: What to know about the water challenge')
        .field
          label.label Description
          .control
            textarea.textarea(placeholder='Description', name="description") Dr. Jennifer Ashton kicks off a month-long "Water Challenge" to look at how drinking more water can affect your health.
        .field
          label.label Link to article
          .control
            input.input(type='text', placeholder='Title', name="url", value='http://abcnews.go.com/GMA/video/water-challenge-52783678')
        .field
          label.label Link to image
          .control
            input.input(type='text', placeholder='Title', name="urlToImage", value='http://a.abcnews.com/images/GMA/180202_gma_ashton_16x9_992.jpg')
        .field
          .control
            button.button.is-primary(type='submit') Submit
