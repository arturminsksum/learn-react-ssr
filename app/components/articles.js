import React, { Component } from 'react';
import Post from './post';

import { requestArticles } from '../actions';
export default class Articles extends Component {
  constructor(props) {
    super(props);
  }

  static getData(dispatch) {
    return dispatch(requestArticles());
  }

  componentDidMount() {
    this.props.requestArticles();
  }

  render() {
    const { articles } = this.props;
    return (
      <section>
        {articles.map((post, index) => <Post item={post} key={index} />)}
      </section>
    );
  }
}
