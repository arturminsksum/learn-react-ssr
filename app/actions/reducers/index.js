import { combineReducers } from 'redux';
import articles from './articles';
import sortArticles from './sortArticles';
const todoApp = combineReducers({
  articles,
  sortArticles,
});
export default articlesApp;
