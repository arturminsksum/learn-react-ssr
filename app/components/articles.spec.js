import React from 'react';

import Articles from './articles';
import articles from '../state.json';
import { requestArticles } from '../actions';

describe('Articles', () => {
  it('should render Articles correctly', () => {
    const wrapper = mount(
      <Articles articles={articles} requestArticles={requestArticles} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
