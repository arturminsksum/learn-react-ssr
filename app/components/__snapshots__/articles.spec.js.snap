// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles should render Articles correctly 1`] = `
<Articles
  articles={
    Array [
      Object {
        "author": "Thuy Ong",
        "description": "Facebook’s new privacy center tool is a response the upcoming General Data Protection Regulation which will be introduced in May",
        "id": "the-verge",
        "publishedAt": "2018-01-29T12:10:18Z",
        "source": "The Verge",
        "title": "Facebook begins privacy push ahead of tough new European law",
        "url": "https://www.theverge.com/2018/1/29/16944304/facebook-privacy-eu-law-general-data-protection-regulation",
      },
      Object {
        "author": "Mix",
        "description": "Earlier in January we wrote about a quirky new trend: cryptocurrency mystery boxes you can order online to get a random selection of digital coins.

CBlocks – the company we spoke ...",
        "id": "the-next-web",
        "publishedAt": "2018-01-29T11:49:06Z",
        "source": "The Next Web",
        "title": "We got a cryptocurrency mystery box – but you probably shouldn't",
        "url": "https://thenextweb.com/hardfork/2018/01/29/cryptocurrency-mystery-box-review/",
      },
      Object {
        "author": "ABC",
        "description": "Ronda Rousey stayed silent as she pointed from the ring toward the WrestleMania logo that hung in the rafters. Without saying a word, the MMA great made clear she was signaling her wrestling debut at WWE's signature event. Rousey wants her WrestleMania moment…",
        "id": "abc-news",
        "publishedAt": "2018-01-29T22:01:09Z",
        "source": "ABC News",
        "title": "Rousey's WWE deal gets ex-UFC star back in bright spotlight",
        "url": "http://abcnews.go.com/Entertainment/wireStory/rouseys-wwe-deal-ufc-star-back-bright-spotlight-52691202",
      },
    ]
  }
  requestArticles={[Function]}
>
  <div>
    <h3
      className="is-size-3"
    >
      Article amount: 
      <span>
        3
      </span>
    </h3>
    <br />
    <section>
      <Component
        item={
          Object {
            "author": "Thuy Ong",
            "description": "Facebook’s new privacy center tool is a response the upcoming General Data Protection Regulation which will be introduced in May",
            "id": "the-verge",
            "publishedAt": "2018-01-29T12:10:18Z",
            "source": "The Verge",
            "title": "Facebook begins privacy push ahead of tough new European law",
            "url": "https://www.theverge.com/2018/1/29/16944304/facebook-privacy-eu-law-general-data-protection-regulation",
          }
        }
        key="0"
      >
        <article
          className="media box"
        >
          <div
            className="media-content"
          >
            <div
              className="content"
            >
              <a
                className="title"
                href="https://www.theverge.com/2018/1/29/16944304/facebook-privacy-eu-law-general-data-protection-regulation"
              >
                Facebook begins privacy push ahead of tough new European law
              </a>
              <p>
                <span>
                  Facebook’s new privacy center tool is a response the upcoming General Data Protection Regulation which will be introduced in May
                </span>
              </p>
              <p>
                <strong>
                  Source: 
                  The Verge
                   
                </strong>
                <small>
                  Author: 
                  Thuy Ong
                   
                </small>
                <small>
                  Time: 
                  2018-01-29T12:10:18Z
                </small>
              </p>
            </div>
          </div>
        </article>
      </Component>
      <Component
        item={
          Object {
            "author": "Mix",
            "description": "Earlier in January we wrote about a quirky new trend: cryptocurrency mystery boxes you can order online to get a random selection of digital coins.

CBlocks – the company we spoke ...",
            "id": "the-next-web",
            "publishedAt": "2018-01-29T11:49:06Z",
            "source": "The Next Web",
            "title": "We got a cryptocurrency mystery box – but you probably shouldn't",
            "url": "https://thenextweb.com/hardfork/2018/01/29/cryptocurrency-mystery-box-review/",
          }
        }
        key="1"
      >
        <article
          className="media box"
        >
          <div
            className="media-content"
          >
            <div
              className="content"
            >
              <a
                className="title"
                href="https://thenextweb.com/hardfork/2018/01/29/cryptocurrency-mystery-box-review/"
              >
                We got a cryptocurrency mystery box – but you probably shouldn't
              </a>
              <p>
                <span>
                  Earlier in January we wrote about a quirky new trend: cryptocurrency mystery boxes you can order online to get a random selection of digital coins.

CBlocks – the company we spoke ...
                </span>
              </p>
              <p>
                <strong>
                  Source: 
                  The Next Web
                   
                </strong>
                <small>
                  Author: 
                  Mix
                   
                </small>
                <small>
                  Time: 
                  2018-01-29T11:49:06Z
                </small>
              </p>
            </div>
          </div>
        </article>
      </Component>
      <Component
        item={
          Object {
            "author": "ABC",
            "description": "Ronda Rousey stayed silent as she pointed from the ring toward the WrestleMania logo that hung in the rafters. Without saying a word, the MMA great made clear she was signaling her wrestling debut at WWE's signature event. Rousey wants her WrestleMania moment…",
            "id": "abc-news",
            "publishedAt": "2018-01-29T22:01:09Z",
            "source": "ABC News",
            "title": "Rousey's WWE deal gets ex-UFC star back in bright spotlight",
            "url": "http://abcnews.go.com/Entertainment/wireStory/rouseys-wwe-deal-ufc-star-back-bright-spotlight-52691202",
          }
        }
        key="2"
      >
        <article
          className="media box"
        >
          <div
            className="media-content"
          >
            <div
              className="content"
            >
              <a
                className="title"
                href="http://abcnews.go.com/Entertainment/wireStory/rouseys-wwe-deal-ufc-star-back-bright-spotlight-52691202"
              >
                Rousey's WWE deal gets ex-UFC star back in bright spotlight
              </a>
              <p>
                <span>
                  Ronda Rousey stayed silent as she pointed from the ring toward the WrestleMania logo that hung in the rafters. Without saying a word, the MMA great made clear she was signaling her wrestling debut at WWE's signature event. Rousey wants her WrestleMania moment…
                </span>
              </p>
              <p>
                <strong>
                  Source: 
                  ABC News
                   
                </strong>
                <small>
                  Author: 
                  ABC
                   
                </small>
                <small>
                  Time: 
                  2018-01-29T22:01:09Z
                </small>
              </p>
            </div>
          </div>
        </article>
      </Component>
    </section>
  </div>
</Articles>
`;
